### PR TITLE
Improve documentation and helm chart generated in driver projects

### LINKED ISSUE #18: Document exception handling
Documentation should explain exception handling in service implementations. In particular:

* how to handle specific exceptions (such as infrastructure already exists) and convert them to FailureDetails objects and in what circumstances this isn't appropriate (if any)?
* how Ignition converts (custom) exceptions in to API responses (including which status codes are applied) and whether this is ever appropriate i.e. should all exceptions be caught by the driver and converted in to FailureDetails objects in the response?

### LINKED ISSUE #25: Documentation should include details of how the framework should be used, preferaby with examples
The documentation for Ignition explains very well how to create a driver, but it does not explain how to actually go about building one.  

The documentation needs to be more detailed in explaining what someone must do to create a driver.  It seems that the documentation at the moment is good at explaining some of the concepts used to create the framework, but does not illustrate to driver creator, what they must do to successfully create a driver.

Things that should be in the documentation are:
  How is a driver to be called - a simple message diagram of the interaction of Brent to the driver would help - or reference the Brent documentation.
  The model of messages that should be used to return values to Brent.
  There is no information or examples around the differences between the lifecycle and infrastructure drivers.

### LINKED ISSUE #26: Steps to build a docker image include incorrect whl directory
The devdocs generated by the `ignition create` command include a step to build the docker image by copying the whl for the app to a `docker/whl` directory when it should be `docker/whls`

### LINKED ISSUE #27: Name used in setup.py of generated driver application should be the module name
The setup.py generated from a `ignition create` command is using the application name as the name field, which leads to whl file names that differ to the module - this is generally considered bad practice.

For example if you create a driver with:
`ignition create "Test Driver" --module-name tdriver`

Then the whl built using setuptools will be called: `Test_Driver-0.0.1-py3-none.whl`
If would be better if this file was: `tdriver-0.0.1-py3-none.whl`

### LINKED ISSUE #28: Correct driver template based on issues from other driver projects
Issues from other driver projects:

- libc and binutils must be added to Docker image for Gunicorn to work
- Dependencies should be fixed to a major version (had issue with a v20.0 of Gunicorn not working, we need to stick with the tested version 19.9)

### PR DESCRIPTION
**Pull request by dvaccarosenna**
Closes #28 
Closes #27 
Closes #26 
Closes #18 
Closes #25

### CODE REVIEW

File: docs/user-guide/api-error-handling.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+# Error Handling
+
+This section assumes you have already created your first driver and started working on your implementation of either the `InfrastructureDriverCapability` or `LifecycleDriverCapability`. It's at this stage you might wonder how to customise the errors thrown by the bootstrapped API services. 
+
+This guide will show you to easily customise the HTTP status code and localizedMessage included in the response by creating a Python exception to be raised in an implementation of a driver capapbility (or your own Service).
```
Comment (sglover) [inline docs/user-guide/api-error-handling.md]: typo: "capapbility"

File: docs/user-guide/creating-a-driver.md [hunk 2]
```diff
@@ -7,6 +7,23 @@ Contents:
 - [Create Command](#create-command)
 - [The Basics of a Driver](#the-basics-of-a-driver)
 
+# Prerequisites
+
+You will need Python3 and Pip to complete this guide.
+
+On some Linux distributions you will need to also installed `python3-dev`. Make sure you install the version which matches your Python version, in most cases the following is suitable:
```
Comment (sglover) [inline docs/user-guide/creating-a-driver.md]: installed -> install

File: docs/user-guide/framework/bootstrap-components/infrastructure.md [hunk 3]
```diff
@@ -23,4 +27,131 @@ Ignition can bootstrap the following services for the Infrastructure API:
...
+
+The InfrastructureService forwards this call down to the `create_infrastructure` method on the `InfrastructureDriver` provided by the user. 
+
+The `InfrastructureDriver` is expected to either:
+- Complete the request by initiating a creation of the infrastructure, returning immediately with an instance of `ignition.model.infrastructure.CreateInfrastructureResponse` which includes an `infrastructure_id` and `request_id` which may be used at a later time to check the status of the infrastructure being created. For example, in the Openstack VIM driver implementation, we return the Stack ID assigned to infrastructure.
+- Raise an Exception if the request is invalid or cannot be accepted. This Exception will end the continuation of this flow and instead result in an error being returned to the client of the InfrastructureApiService.
+
+On receival of a CreateInfrastructureResponse, the InfrastructureService will inform the InfrastructureTaskMonitoringService that it should monitor the completion of the infrastructure creation, using the `infrastructure_id` and `request_id` on the response.
```
Comment (sglover) [inline docs/user-guide/framework/bootstrap-components/infrastructure.md]: receival -> receipt?

File: docs/user-guide/framework/bootstrap-components/infrastructure.md [hunk 4]
```diff
@@ -23,4 +27,131 @@ Ignition can bootstrap the following services for the Infrastructure API:
...
+
+The InfrastructureTaskMonitoringService will periodically, using the JobQueueService (see [job queue](./job_queue)), check if infrastructure has been created. It will do this by polling the `get_infrastructure_task` method of the `InfrastructureDriver`
```
Comment (sglover) [inline docs/user-guide/framework/bootstrap-components/infrastructure.md]: Is it technically "polling" or just "calling" (polling may imply that it's sat there in a poll loop, perhaps)?

File: docs/user-guide/framework/bootstrap-components/infrastructure.md [hunk 5]
```diff
@@ -23,4 +27,131 @@ Ignition can bootstrap the following services for the Infrastructure API:
...
+        :return: an ignition.model.infrastructure.FindInfrastructureResponse
+        """
+        pass
+```
+
+There are several methods required on any Service implementing this Capability, in order to serve Create/Delete/Find infrastructure requests. The following sections describe the flow of an incoming request, outlines the role of Ignition and how your driver implementation is used.
+
+## On Create
+
+The InfrastructureApiService handles the incoming HTTP request, parsing the headers and body before making a call to the InfrastructureService with the expected arguments.
```
Comment (sglover) [inline docs/user-guide/framework/bootstrap-components/infrastructure.md]: Probably a pain to do (if even possible), but is it worth adding a sequence diagram here?

File: docs/user-guide/framework/bootstrap-components/infrastructure.md [hunk 6]
```diff
@@ -23,4 +27,131 @@ Ignition can bootstrap the following services for the Infrastructure API:
...
+| Failure Code | Description |
+| --- | --- |
+| RESOURCE_ALREADY_EXISTS | An element of the Infrastructure cannot be created |
+| INFRASTRUCTURE_ERROR | There was an error creating the Infrastructure |
+| INSUFFICIENT_CAPACITY | There is not enough capacity in the VIM to create this Infrastructure |
+| INTERNAL_ERROR | For all other types of errors |
+
+## On Delete
+
+The flow of a delete infrastructure request is the same as a create - instead of `create_infrastructure` the `delete_infrastructure` method of the user provided `InfrastructureDriver` is called instead and a `ignition.model.infrastructure.DeleteInfrastructureResponse` is expected.
```
Comment (sglover) [inline docs/user-guide/framework/bootstrap-components/infrastructure.md]: superfluous "instead"

File: docs/user-guide/framework/bootstrap-components/infrastructure.md [hunk 7]
```diff
@@ -23,4 +27,131 @@ Ignition can bootstrap the following services for the Infrastructure API:
...
+
+The `InfrastructureDriver` is expected to complete the request and return an instance of a `ignition.model.infrastructure.FindInfrastructureResponse` with details of whether the infrastructure could be found or not. 
+
+The InfrastructureService takes this result and returns it to the InfrastructureApiService so it may be converted to a HTTP response and returned to the original client.
+
+## Errors
+
+You are free to raise any Python errors in your code to indicate failure of a request. To customise the response returned to the API client on error, see [error handling](../../api-error-handling.md).
+
+However, please note, in the `infrastructure` module of Ignition there already some error types you may use:
```
Comment (sglover) [inline docs/user-guide/framework/bootstrap-components/infrastructure.md]: minor: "there" -> "there are"

File: docs/user-guide/framework/bootstrap-components/job_queue.md [hunk 8]
```diff
@@ -1,8 +1,8 @@
 # Job Queue
 
-A Job Queue is a core feature required by many of the Services expected in a VIM or VNFC driver. As a result, all of the Services listed below are enabled by default when using the `ignition.boot.api.build_vim_driver` or `ignition.boot.api.build_vnfc_driver` method. 
+A Job Queue is a core feature required by many of the Services expected in a VIM or Lifecycle driver. As a result, all of the Services listed below are enabled by default when using the `ignition.boot.api.build_vim_driver` or `ignition.boot.api.build_lifecycle_driver` method. 
 
-In Ignition a job queue is implemented using a message bus, rather than in-memory, to maintain high availability. Jobs are queued on the message bus, each instance of the VIM/VNFC driver will take jobs from the queue, then either complete them or requeue them by posting it back to the bus.
+In Ignition a job queue is implemented using a message bus, rather than in-memory, to maintain high availability. Jobs are queued on the message bus, each instance of the VIM/Lifecycle driver will take jobs from the queue, then either complete them or requeue them by posting it back to the bus.
```
Comment (sglover) [inline docs/user-guide/framework/bootstrap-components/job_queue.md]: minor: "by posting it" -> "by posting them"

File: ignition/openapi/vim_infrastructure.yaml [hunk 9]
```diff
@@ -1,113 +1,190 @@
...
+    description: Infrastructure APIs
 paths:
   /create:
     post:
       tags:
         - infrastructure
       summary: Create infrastructure
       description: >-
-        Create infrastructure
+        Create infrastructure described by a Template string in the given Deployment Location. If accepted, a response should be returned immediately with an infrastructureId, to refer to the infrastructured being created, and a requestId, to refer to the creation process.
```
Comment (sglover) [inline ignition/openapi/vim_infrastructure.yaml]: minor: "infrastructured" -> "infrastructure"

File: ignition/openapi/vim_infrastructure.yaml [hunk 10]
```diff
@@ -1,113 +1,190 @@
...
+                example:
+                  $ref: "#/components/examples/ExampleQueryTaskErrorResponse"
   /delete:
     post:
       tags:
         - infrastructure
       summary: Deletes infrastructure
-      description: Deletes infrastructure
+      description: >-
+        Deletes infrastructure previously created in the given Deployment Location. If the request is accepted, a response should be returned immediately with a requestId to refer to the deletion process. The completion of the request (regardless of whether it was successful or not) should be returned via Kafka
```
Comment (sglover) [inline ignition/openapi/vim_infrastructure.yaml]: "should be returned via Kafka" - perhaps indicate that this is handled by Ignition by virtue of the fact that the driver returns a response in its get method?

File: ignition/templates/driver_template/devdocs/helm-chart.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    - uWSGI container used by your application
+    - Pod affinity/anti-affinity
+    - Node affinity/anti/affinity
+    - NodePort
+    - Ingress Host
+- Necessary Deployment labels so logs from the driver may be viewed in the dashboard provided with LM (Kibana)
+
+# Build Helm Chart
+
+To build the Helm chart you will need to install Helm and initialised it (e.g. `helm init --client-only` on your local machine).
```
Comment (sglover) [inline ignition/templates/driver_template/devdocs/helm-chart.md]: "initialised" -> "initialise"

File: ignition/templates/driver_template/devdocs/helm-chart.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+On Node:
+```
+docker load -i /home/<user>/{(docker.name)}-img.tar
+```
+
+## Use Docker Registry
+
+In a custom **Helm values file** add the following:
+
```
Comment (sglover) [inline ignition/templates/driver_template/devdocs/helm-chart.md]: Worth mentioning here that "docker-daemon" insecure registries needs to be configured appropriately for this to work?

File: ignition/templates/driver_template/devdocs/python-app.md [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+The commands `{(app.module_name)}-dev`, `{(app.module_name)}`, `{(app.module_name)}-gunicorn` and `{(app.module_name)}-uswgi` will now be avaiable from the command line.
+
+## Configuration
+
+Ignition loads configuration properties from any sources provided to the application builder. By default this driver is configured to load properties from:
+
+- `{(app.module_name)}/config/default_config.yml` - this is the default configuration file, bundled into the distributed version of your application
+- `{(app.module_name)}_config.yml` - Ignition will search for a configuration file with this name at the directory the application has been started from (ignored if not found)
+- `/var/{( app.module_name )}/{( app.module_name )}_config.yml` - this configuration file is only used in the Helm chart installation. Ignition will search for a configration file at this path (ignored if not found)
```
Comment (sglover) [inline ignition/templates/driver_template/devdocs/python-app.md]: "configration" -> "configuration"

File: docs/user-guide/framework/bootstrap-components/infrastructure.md (hunk 5 above)
Comment (dvaccarosenna) [reply docs/user-guide/framework/bootstrap-components/infrastructure.md]: I agree, just ran out of time really to add it. I'll raise that as a separate issue (https://github.com/accanto-systems/ignition/issues/33) so we can do it when time allows.

File: ignition/templates/lifecycle_additions/{(app.module_name)}/service/lifecycle.py [hunk 14]
```diff
@@ -8,7 +8,7 @@ def execute_lifecycle(self, lifecycle_name, lifecycle_scripts_tree, system_prope
         """
         Execute a lifecycle transition/operation for a Resource.
         This method should return immediate response of the request being accepted,
-        it is expected that the LifecycleService will poll get_lifecycle_execution on this driver to determine when the request has complete (or devise your own method).
+        the LifecycleService will poll get_lifecycle_execution on this driver to determine when the request has complete (or devise your own method).
```
Comment (sglover) [inline ignition/templates/lifecycle_additions/{(app.module_name)}/service/lifecycle.py]: "has complete" -> "has completed"

File: ignition/templates/vim_additions/{(app.module_name)}/service/infrastructure.py [hunk 15]
```diff
@@ -8,7 +8,7 @@ def create_infrastructure(self, template, template_type, inputs, deployment_loca
         """
         Initiates a request to create infrastructure based on a TOSCA template.
         This method should return immediate response of the request being accepted,
-        it is expected that the InfrastructureService will poll get_infrastructure_task on this driver to determine when the request has complete.
+        the InfrastructureService will poll get_infrastructure_task on this driver to determine when the request has complete.
```
Comment (sglover) [inline ignition/templates/vim_additions/{(app.module_name)}/service/infrastructure.py]: complete -> "completed"

File: tests/unit/templates/expected_lifecycle_driver/devdocs/helm-chart.md [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+    - uWSGI container used by your application
+    - Pod affinity/anti-affinity
+    - Node affinity/anti/affinity
+    - NodePort
+    - Ingress Host
+- Necessary Deployment labels so logs from the driver may be viewed in the dashboard provided with LM (Kibana)
+
+# Build Helm Chart
+
+To build the Helm chart you will need to install Helm and initialised it (e.g. `helm init --client-only` on your local machine).
```
Comment (sglover) [inline tests/unit/templates/expected_lifecycle_driver/devdocs/helm-chart.md]: "initialised" -> "initialise"

**Pull request merged**